fault-proof/src/prover.rs

```diff
@@ -0,0 +1,580 @@
+            .request_async()
+            .await?;
+
+        tracing::info!(proof_id = %proof_id, "Range proof request submitted");
+        Ok(proof_id)
+    }
+
+    /// Submit an aggregation proof request to the network.
+    async fn request_agg_proof(&self, stdin: &SP1Stdin) -> Result<ProofId> {
+        let proof_id = self
+            .prover
+            .prove(&self.keys.agg_pk, stdin)
+            .mode(self.config.agg_proof_mode)
+            .strategy(self.config.agg_proof_strategy)
+            .timeout(Duration::from_secs(self.config.timeout))
+            .min_auction_period(self.config.min_auction_period)
+            .max_price_per_pgu(self.config.max_price_per_pgu)
+            .cycle_limit(self.config.agg_cycle_limit)
+            .gas_limit(self.config.agg_gas_limit)
+            .whitelist(self.config.whitelist.clone())
+            .request_async()
+            .await?;
+
+        tracing::info!(proof_id = %proof_id, "Aggregation proof request submitted");
+        Ok(proof_id)
+    }
+
+    /// Wait for a proof to be fulfilled by polling the network.
+    ///
+    /// Timeout behavior:
+    /// - **Proving timeout** (`config.timeout`): Overall maximum wait time from start.
+    /// - **Network call timeout** (`config.network_calls_timeout`): Per-call timeout; retries on
+    ///   failure.
+    /// - **Auction timeout** (`config.auction_timeout`): Cancels if no prover picks up the request.
+    /// - **Server deadline** (`status.deadline()`): Server-side proving deadline.
+    async fn wait_for_proof(&self, proof_id: ProofId) -> Result<SP1ProofWithPublicValues> {
+        let start_time = std::time::Instant::now();
+        let proving_timeout = Duration::from_secs(self.config.timeout);
+        let is_mainnet = self.network_mode == NetworkMode::Mainnet;
+
+        loop {
+            // Proving timeout - ensures we don't wait forever if network calls keep failing.
+            if let ProvingTimeout::Exceeded { elapsed_secs } =
+                check_timeout(start_time.elapsed(), proving_timeout)
+            {
+                tracing::warn!(
+                    proof_id = %proof_id,
+                    elapsed_secs,
+                    timeout_secs = self.config.timeout,
+                    "proving timeout exceeded"
+                );
+                ProposerGauge::ProvingTimeoutError.increment(1.0);
+                bail!(
+                    "Proving timeout: proof_id={}, elapsed={}s, timeout={}s \
+                    (consider reducing workload: increase RANGE_SPLIT_COUNT or reduce PROPOSAL_INTERVAL_IN_BLOCKS)",
+                    proof_id,
+                    elapsed_secs,
+                    self.config.timeout
+                );
+            }
+
+            // Get proof status - retry on transient failures.
+            let (status, proof) = match self
+                .network_call_with_timeout(
+                    self.prover.get_proof_status(proof_id),
```

Comment:
Shouldn't get_proof_request be enough? Why do we need this call?

Reply:
`get_proof_status` returns the fulfillment status and proof, but doesn't include the `created_at` timestamp. We need `get_proof_request` to get the `created_at` field, which is required to calculate whether the auction timeout has been exceeded (i.e., whether `current_time > created_at + auction_timeout` when the request is still in `Requested` state).

This is something looks like should be improved on the SDK side. Ideally `get_proof_status` would include `created_at` so we don't need two separate API calls per poll iteration.